Making ExternalSpillableMap generic for any datatype

n3nash opened the pull request:
- Introduced concept of converters to be able to serde generic datatype for SpillableMap
	- Fixed/Added configs to Hoodie Configs
	- Changed HoodieMergeHandle to start using SpillableMap

n3nash commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieWriteConfig.java:
```diff
@@ -64,6 +64,9 @@
   private static final String DEFAULT_FINALIZE_WRITE_PARALLELISM = DEFAULT_PARALLELISM;
+  private static final String MAX_SIZE_IN_MEMORY_FOR_MERGE_IN_BYTES_PROP = "hoodie.merge.spill.threshold";
+  // Default max memory size during hash-merge, excess spills to disk
+  private static final String DEFAULT_MAX_SIZE_IN_MEMORY_FOR_MERGE_IN_BYTES = String.valueOf(1024*1024*1024L); //1GB
```
Make it fraction instead of absolute value.

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java:
```diff
@@ -156,7 +159,8 @@ public R put(T key, R value) {
         // Naive approach for now
         if (estimatedPayloadSize == 0) {
-          this.estimatedPayloadSize = SpillableMapUtils.computePayloadSize(value, schema);
+          this.estimatedPayloadSize = SpillableMapUtils.computePayloadSize(key, this.keyConverter) +
```
use SizeEstimator from spark.

n3nash commented:
@vinothchandar please take a pass at this when you get a chance.

vinothchandar commented on docs/configurations.md:
```diff
@@ -23,7 +23,9 @@ summary: "Here we list all possible configurations and what they mean"
         <span style="color:grey">Should HoodieWriteClient assume the data is partitioned by dates, i.e three levels from base path. This is a stop-gap to support tables created by versions < 0.3.1. Will be removed eventually </span>
-    
+    - [withMaxMemoryFractionPerPartitionMerge](#withMaxMemoryFractionPerPartitionMerge) (maxMemoryFractionPerPartitionMerge = 0.6) <br/>
+    <span style="color:grey">This fraction is multiplied with the spark.memory.fraction to get a final fraction of heap space to use during merge </span>
```
We should also document how this interplays with `spark.memory.storageFraction` i.e this + that <= 1.0, otherwise things will OOM

n3nash replied:
Actually, this is just defining what fraction of spark.memory.fraction is used for in-memory map in MergeHandle. So if the client uses sparkconfigs incorrectly, spark itself will cause issues before it hits the mergehandle code..

vinothchandar replied:
I still think its worth documenting..
http://spark.apache.org/docs/2.1.1/tuning.html#memory-management-overview

 Usually, spark.memory.fraction - spark.memory.storageFraction is left for internal datastructures of spark. Heap - spark.memory.fraction is left for user data structures - in this case hoodie externalspillablemap.. 

I dont actually follow 
  - Why we multiply this fraction against spark fraction instead of total heap space. By definition, we cannot dip into the spark fraction
  - How can the spark app be stable is spark fraction is 0.6 and this also takes 0.6 * 0.6 of heap, which will bring total usage to 0.6 + 0.36 = 0.96 or 96% full heap, which will keep gcing back to back..

n3nash replied:
The definitions in this link : https://spark.apache.org/docs/latest/configuration.html and the one you pasted (which is the description link in the above link) are a little misleading. I think there was an oversight on my part for choosing to multiple with spark.memory.fraction. Instead, I should multiply with 1 - spark.memory.fraction. It's more clear here : https://0x0fff.com/spark-memory-management/. 
I don't want to rely on heap space to calculate the spillablemap size since the heap is actually not the real user memory left by spark memory model. 
So my thought process is, executor.memory = heap. So as you increase executor.memory you increase the heap. We want our spillablemap memory size to grow accordingly, hence user.available.memory = executor.memory * (1 - spark.memory.fraction)
spillable.available.memory = user.available.memory*merge.memory.fraction.

@vinothchandar

vinothchandar replied:
`spillable.available.memory = user.available.memory*merge.memory.fraction` sure, this is what I was getting at in the comments above. This should be okay.

vinothchandar replied:
@n3nash lets reflect this discussion in the docs above?  Multiplying by spark fraction is definitely misleading

n3nash replied:
done

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2016 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
```
2018

n3nash replied:
done

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,101 @@
+  public static final String MAX_MEMORY_FRACTION_FOR_MERGE_PROP = "hoodie.merge.memory.fraction";
+  // Default max memory fraction during hash-merge, excess spills to disk
+  public static final String DEFAULT_MAX_MEMORY_FRACTION_FOR_MERGE = String.valueOf(0.6);
+  public static final String SPARK_EXECUTOR_MEMORY_PROP = "spark.executor.memory";
```
Why expose these variables from here? Can we just return the computed memory limit from the config class instead of passing all this to IOHandle level?

n3nash replied:
Yeah, I wanted to do that but sent out the diff to get a first revision anyways..

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,101 @@
+      .valueOf(1024 * 1024 * 1024L); // 1GB
+
+  // TODO(NA):convert this to fraction, use either Runtime.freeMemory() or spark.executor.memory fractions depending on execution
+  public static final String MAX_SIZE_IN_MEMORY_PER_COMPACTION_IN_BYTES_PROP = "hoodie.compaction.spill.threshold";
```
I feel this should stay in `CompactionConfig` or be renamed.. High level, its debatable if we need a class for these two configs.. I am with you on bringing all the memory management related stuff into a place, but this list feels too small..  thoughts?

n3nash replied:
I think separation makes sense right now.

vinothchandar replied:
lets chat more.. Not in favor of introducing new classes, while deferring their real usage down the line.. Happy to consider this if we can group more memory related params.. are there others that can live here?

n3nash replied:
This is not exactly deferring their usage down the line, there's more than 1 memory config being used here and adding them to the HoodieWriteConfig or HoodieCompactionConfig seems like a wrong place. But if you feel strongly about not adding another class, I can move these configs to HoodieWriteConfig.

vinothchandar replied:
I dont think we are going to agree. I ll bite.. but can we rename `hoodie.compaction.spill.threshold` to `hoodie.memory.compaction.spill.threshold` to be consistent?

n3nash replied:
Renamed it!

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,101 @@
+  public static final String DEFAULT_MAX_MEMORY_FRACTION_FOR_MERGE = String.valueOf(0.6);
+  public static final String SPARK_EXECUTOR_MEMORY_PROP = "spark.executor.memory";
+  public static final String SPARK_EXECUTOR_MEMORY_FRACTION_PROP = "spark.memory.fraction";
+  public static final String DEFAULT_SPARK_EXECUTOR_MEMORY_FRACTION = "0.6";
```
spark default is 0.5, are we bumping this up to 0.6? Can we get by without overriding spark default inside hudi ?

n3nash replied:
It's not overriding. So there is a disconnect in what the spark docs say as default : https://spark.apache.org/docs/latest/configuration.html

and in the code : https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkConf.scala#L507
I plan to open a ticket on Spark for this, for now, going with what is in the docs...

vinothchandar replied:
Sure, still why are we defining spark configs defaults here? Spark config is something I totally want to leave it upto user while providing guidelines alone

n3nash replied:
Just explained the reason for this in another comment, hope that suffices.

vinothchandar replied:
Its orthogonal. I still don't understand why we are definiing spark config values in this file.. Can we for e.g decide dynamically based on values in `SparkConf`.
i.e `SPARK_EXECUTOR_MEMORY_PROP` , `SPARK_EXECUTOR_MEMORY_FRACTION_PROP` should not be defined in a HoodieConfig file right?  What am I missing?

n3nash replied:
So, there are 2 ways to determine the spillable.memory.memory. We just let the user pass a fraction of available.memory that can be used to calculate this. In such a case, we want to get the `spark.executor.memory` and the `spark.memory.fraction`. To get these values from `SparkConf`, we need to know which properties to look for. Hence, these PROPs. The other option is to let the user figure out the spillable.memory and let the user correctly size the  spillable.memory when executory memory is increased or decreased, in which case we don't need the props or the calculation..

vinothchandar replied:
> To get these values from SparkConf, we need to know which properties to look for 

is there constant inside Spark we can use without defining them here? thats the crux of what I am asking.. lets chat f2f, something is getting lost in translation

n3nash replied:
I see it hardcoded everywhere in  the code, across tests and src's. Even in https://github.com/apache/spark/blob/5fa438471110afbf4e2174df449ac79e292501f8/core/src/main/scala/org/apache/spark/SparkConf.scala#L502 and https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkContext.scala#L465

vinothchandar replied:
The two constants are used in just the one method.. Can we move them as local final variables there.. I don't want to mix them with other hoodie level properties we declare here.. 

Also can we get away without defining the defaults here? i.e assume these values are passed down from app or spark installation?

n3nash replied:
Move the variables as final. I think we cannot assume that these values are passed down from the app or spark installation since one can instantiate a HoodieWriteConfig even before a SparkContext or SparkConf is created. We could move this method to where it was earlier, in the HoodieIOHandle/ HoodieMergeHandle, then we can be sure that the SparkContext / SparkConf has been created. WDTY ? In the meantime, I can check more on this

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,101 @@
+  // This makes it easier to scale this value as one increases the spark.executor.memory
+  public static final String MAX_MEMORY_FRACTION_FOR_MERGE_PROP = "hoodie.merge.memory.fraction";
+  // Default max memory fraction during hash-merge, excess spills to disk
+  public static final String DEFAULT_MAX_MEMORY_FRACTION_FOR_MERGE = String.valueOf(0.6);
```
0.6 (merge fraction) + 0.6 (spark storage fraction) > 1.0. won't this be a problem?

n3nash replied:
I'm not introducing new fractions : https://spark.apache.org/docs/latest/configuration.html

Please read the document.

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,101 @@
+  public static final String SPARK_EXECUTOR_MEMORY_PROP = "spark.executor.memory";
+  public static final String SPARK_EXECUTOR_MEMORY_FRACTION_PROP = "spark.memory.fraction";
+  public static final String DEFAULT_SPARK_EXECUTOR_MEMORY_FRACTION = "0.6";
+  public static final String DEFAULT_SPARK_EXECUTOR_MEMORY = "1024"; // in MB
```
rename : `DEFAULT_SPARK_EXECUTOR_MEMORY_MB`

n3nash replied:
done

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieWriteConfig.java:
```diff
@@ -68,6 +68,7 @@
   private static final String FINALIZE_WRITE_PARALLELISM = "hoodie.finalize.write.parallelism";
   private static final String DEFAULT_FINALIZE_WRITE_PARALLELISM = DEFAULT_PARALLELISM;
 
+
```
\- extra line

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieWriteConfig.java:
```diff
@@ -108,7 +109,8 @@ public int getUpsertShuffleParallelism() {
 
   public int getWriteBufferLimitBytes() {
-    return Integer.parseInt(props.getProperty(WRITE_BUFFER_LIMIT_BYTES, DEFAULT_WRITE_BUFFER_LIMIT_BYTES));
+    return Integer
```
Can we avoid additional formatting changes across the diff and trim it down to only real changes?

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieWriteConfig.java:
```diff
@@ -338,6 +342,39 @@ public String getGraphiteMetricPrefix() {
+    return HoodieMemoryConfig.DEFAULT_SPARK_EXECUTOR_MEMORY_FRACTION;
+  }
+
+  public static String getDefaultSparkExecutorMemory() {
```
why explicit getters for already public static constants?  That said, I recommend we localize all this to the config class

n3nash replied:
So the way our HoodieConfigs are setup, we expose getters from a single HoodieWriteConfig and then depend on the constants from the specific configs, hence those constants need to be public so they can be accesses in HoodieWriteConfig and I want one way for client to access them (via getters) rather than using the public constants, this isn't new, the same applied already for compaction, metrics...

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/io/HoodieIOHandle.java:
```diff
@@ -65,10 +69,12 @@ public Path makeNewPath(String partitionPath, int taskPartitionId, String fileNa
   }
 
-  public Path makeTempPath(String partitionPath, int taskPartitionId, String fileName, int stageId, long taskAttemptId) {
+  public Path makeTempPath(String partitionPath, int taskPartitionId, String fileName, int stageId,
```
May I suggest having your autosave option to only format changed lines? (if thats doable)

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/io/HoodieIOHandle.java:
```diff
@@ -104,4 +110,29 @@ public Schema getSchema() {
   }
+
+  // Dynamic calculation of maxMemory to use for merge
+  protected long getMaxMemoryAllowedForMerge() {
```
javadoc style comments .. this method probably deserves more documentation also

vinothchandar replied:
Can we move this to the config class itself?

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/io/HoodieIOHandle.java:
```diff
@@ -104,4 +110,29 @@ public Schema getSchema() {
+      long executorMemoryInBytes = Long.valueOf(
+          Utils.memoryStringToMb(SparkEnv.get().conf().get(config.getSparkExecutorMemoryProp(),
+              config.getDefaultSparkExecutorMemory())) * 1024
+              * 1024L); // 1 GB is the default conf used by Spark, look at SparkContext.scala
```
how do we intend to keep up with the default.. is there a equivalent way to do this dynamically as below?

n3nash replied:
I checked earlier, it's hardcoded in Spark..:( 

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkContext.scala#L465

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/io/HoodieIOHandle.java:
```diff
@@ -104,4 +110,29 @@ public Schema getSchema() {
+          "MemoryFraction of executor memory allowed for merge => " + maxMemoryFractionForMerge);
+      long maxMemoryForMerge = (long) Math
+          .floor(executorMemoryInBytes * memoryFraction * maxMemoryFractionForMerge);
+      logger.info("MaxMemoryInBytes allowed for merge => " + maxMemoryForMerge);
```
nit: can you group these log statements into a single one, that prints all different values? it will definitely be helpful while debugging to have this math in a single log line

n3nash replied:
done

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java:
```diff
@@ -156,9 +164,21 @@ private String init(String fileId, Iterator<HoodieRecord<T>> newRecordsItr) {
     return partitionPath;
   }
 
+  private void logExtra() {
```
same comment. can we do a single log line? and get rid of this method?

n3nash replied:
yup

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/converter/DefaultKeyConverter.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * A default converter implementation for String type of payload key
+ */
+public class DefaultKeyConverter implements Converter<String> {
```
can we rename to `StringConverter`?  this way we can use the class, as long its a string..

n3nash replied:
good point, done.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/converter/DefaultKeyConverter.java:
```diff
@@ -0,0 +1,40 @@
+
+  @Override
+  public byte[] getBytes(String s) {
+    return s.getBytes();
```
need to pay attention to the charset for consistency.. please have it use `utf-8` explcitly..

n3nash replied:
good catch, done.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/converter/DefaultValueConverter.java:
```diff
@@ -0,0 +1,135 @@
+/**
+ * A default converter implementation for HoodieRecord
+ */
+public class DefaultValueConverter<V> implements
```
rename to `HoodieRecordConvertor` ?

n3nash replied:
done

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/converter/DefaultValueConverter.java:
```diff
@@ -0,0 +1,135 @@
+      Field[] fields = clazz.getDeclaredFields();
+      Optional<Field> fieldWithSchema = Arrays.stream(fields)
+          .filter(field -> {
+            if (field.getType() == Schema.class) {
```
can we merge these if statements?

n3nash replied:
done

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/converter/DefaultValueConverter.java:
```diff
@@ -0,0 +1,135 @@
+    // TODO : Find a generic way to figure out the true size of the record in any scenario
+    long sizeOfRecord = ObjectSizeCalculator.getObjectSize(hoodieRecord);
+    long sizeOfSchema = ObjectSizeCalculator.getObjectSize(schema);
+    log.info("SizeOfRecord => " + sizeOfRecord);
```
single log statement.. also should this be info?

n3nash replied:
Yeah, INFO seems fine, it's only logged once right now since the size estimate is done once.

vinothchandar commented on hoodie-common/src/test/java/com/uber/hoodie/common/model/TestAvroPayload.java:
```diff
@@ -0,0 +1,58 @@
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.IndexedRecord;
+
+public class TestAvroPayload implements HoodieRecordPayload {
```
can we extend or reuse an existing Payload class?

n3nash replied:
Unfortunately not, this is a special payload implementation which holds only bytes.

vinothchandar replied:
rename : AvroBinaryTestPayload

n3nash replied:
done

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/DiskBasedMap.java:
```diff
@@ -48,10 +47,8 @@
+final public class DiskBasedMap<T, R> implements Map<T, R> {
 
   // Stores the key and corresponding value's latest metadata spilled to disk
   final private Map<T, ValueMetadata> inMemoryMetadataOfSpilledData;
```
done

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java:
```diff
@@ -33,88 +32,79 @@
- * writes.
- * @param <T>
- * @param <R>
+ * An external map that spills content to disk when there is insufficient space for it to grow. <p>
```
It's resulting from the fact that earlier checkins were not following google style, now every check in does.

vinothchandar replied:
okay.. the checkstyle stuff will hopefully fix this for good.

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java:
```diff
@@ -156,37 +146,38 @@ public R put(T key, R value) {
       return value;
-    } catch(IOException io) {
-      throw new HoodieIOException("Unable to estimate size of payload", io);
+    } catch (Exception io) {
```
removed exception

n3nash commented on hoodie-common/src/test/java/com/uber/hoodie/common/util/collection/TestDiskBasedMap.java:
```diff
@@ -92,16 +131,58 @@ public void testSimpleUpsert() throws IOException, URISyntaxException {
   }
+
+  @Test
+  public void testSizeEstimator() throws IOException, URISyntaxException {
```
It's not called in a loop, it's just called once.

vinothchandar replied:
okay sg.

n3nash commented:
@vinothchandar I've addressed your comments and left some more comments. Please take a pass. I'm going to do some microbenchmarking on the sizeEstimator in the meantime and report back

n3nash commented:
The sizeEstimator, which is using twitter ObjectSizeCalculator, shows a min of 95 ms and max of 180 ms as below : 

18/03/22 07:18:49 INFO collection.ExternalSpillableMap: Estimated Payload size => 1049008 Time taken 126
18/03/22 07:18:49 INFO collection.ExternalSpillableMap: Estimated Payload size => 1037889 Time taken 180
18/03/22 07:18:49 INFO collection.ExternalSpillableMap: Estimated Payload size => 1038006 Time taken 95

For now, I will live with calculating the payload size once at the beginning, if that shows problems will invest in a config based every N records size estimate.

n3nash commented:
@vinothchandar I did a little more testing and realized it's really difficult to come up with a concrete object size for an entry in the hashmap in the JVM since objects can be shared etc. I'm thinking if spillableMap should be made config based for now ? Need to spend more time to understand how to do object sizing so that records don't spill to disk when they are not needed to. WDYT ?

vinothchandar commented:
>>I'm thinking if spillableMap should be made config based for now
Meaning, you want to control if we use spillable map or totally in-memory hashmap? 

>> I did a little more testing and realized it's really difficult to come up with a concrete object size for an entry in the hashmap in the JVM since objects can be shared etc.
Can we especially hard if we hold avro records with a shared schema object... Is this a new problem with the twitter estimator? we were using Spark's estimator in other places right.. 

Can you provide me more context in to whats going wrong

n3nash commented:
Yeah, so say we have a Map<HoodieRecords>. In most scenarios, what ends up happening is a HoodieRecord holds a payload which might in turn hold a GenericRecord. The size of such a HoodieRecord is 1049008. When I drill into the HoodieRecord, it turns out that schema is around 100000 and the rest is 49008. If you estimate the size of hoodie record, we would assume it to be 1049008. But actually, the schema object is shared amongst all payloads of all HoodieRecords, so ideally, the real payload size is 49008. I've been able to address this issue in the `HoodieRecordConverter` by performing some Class field look ups, you might have noticed this during your review, and subtracting the sizeOfSchema from the sizeOfRecord to get the actual HoodieRecord size in the heap, which is 49008. So all good uptill now but still a little hacky. During some more offline tests, I realized that for a more complex payload, which hold nested schemas, I cannot do any such optimizations an hence end up over-estimating the size of the record. The downside of overestimating is that now we will spillToDisk much sooner than we would have if we were using a regular HashMap<>.

vinothchandar commented:
I think holding it serialized could address some of these concerns? Taking a page out of Spark's book, this is why serialized in-memory storage is better/more compact at more CPU cost.. is that an option?

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,136 @@
+      return this;
+    }
+
+    // Dynamic calculation of maxMemory to use for merge
```
java doc style comments

n3nash replied:
done

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,136 @@
+
+  // This fraction is multiplied with the spark.memory.fraction to get a final fraction of heap space to use during merge
+  // This makes it easier to scale this value as one increases the spark.executor.memory
+  public static final String MAX_MEMORY_FRACTION_FOR_MERGE_PROP = "hoodie.merge.memory.fraction";
```
rename to `hoodie.memory.merge.fraction` in sync with how other configs are named?

n3nash replied:
done

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -0,0 +1,136 @@
+  // This fraction is multiplied with the spark.memory.fraction to get a final fraction of heap space to use during merge
+  // This makes it easier to scale this value as one increases the spark.executor.memory
+  public static final String MAX_MEMORY_FRACTION_FOR_MERGE_PROP = "hoodie.merge.memory.fraction";
+  public static final String MAX_MEMORY_FOR_MERGE_PROP = "hoodie.max.merge.memory";
```
same here.. different name..

n3nash replied:
done

n3nash commented:
@vinothchandar Found out a better (and simple) way to accurately size an entry in a map and works well for any type of payload. I've incorporated that and your comments, please take a pass.

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java:
```diff
@@ -142,51 +136,60 @@ public boolean containsValue(Object value) {
+            keyConverter.sizeEstimate(key) + valueConverter.sizeEstimate(value);
+        log.info("Estimated Payload size => " + estimatedPayloadSize);
+      }
+      else if(shouldEstimatePayloadSize &&
```
@vinothchandar Simple handling of sizes.

vinothchandar replied:
how does estimating the size of entire map alleviate the issues you mentioned before?

n3nash replied:
The inherent problem lies in the fact that an entry in the Map by itself is not a good gauge of how large the record is in the heap (as discussed before mainly due to shared objects). What essentially is required is to estimate the size of `a record` given `N records` collectively, that way the size of the shared object is `amortized` over N records and we get as close to the actual payload size as possible. Here, estimating the size of the map does that for us. It's still not dead accurate but comes very close, chose the number `100` because tried with a couple different payloads, nested and non-nested with shared objects and the size of the shared object get's amortized. We start with estimating the payload size for one record first, use that till we reach 100 records (since we don't want to OOM given a small memory setting and large record) and then update the payload estimate. We could choose a higher number say `1000` at which we want to re-estimate which might give use better amortization but the fear is that because of overestimation we may already start spilling to disk by then. 
The other option is to keep updating the estimate logarithmically, but my fear is for a huge HashMap, I'm not sure how the `ObjectSizeEstimator` performs in terms of CPU and Memory usage.

vinothchandar replied:
lets do a follow on task to fix this more systematically.. if 100 works for now, may be we can keep it as it is. 

IIUC you are saying by estimating at a higher point in the object tree (graph), the sizing is more accurate? If so, then that makes sense..

n3nash replied:
Yes, that's what I'm saying.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java:
```diff
@@ -142,51 +136,60 @@ public boolean containsValue(Object value) {
-      return value;
-    } catch(IOException io) {
-      throw new HoodieIOException("Unable to estimate size of payload", io);
+      if (!inMemoryMap.containsKey(key)) {
```
I think we are assuming records are of the same size here and updates need not adjust for size..  worthy to leave a TODO here to revisit..

n3nash replied:
Yes, we are assuming that, done.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java:
```diff
@@ -33,88 +33,82 @@
+  // Find the actual estimated payload size after inserting N records
+  final private static int NUMBER_OF_RECORDS_TO_ESTIMATE_PAYLOAD_SIZE = 100;
+  // Flag to determine whether to stop re-estimating payload size
+  private boolean shouldEstimatePayloadSize = true;
```
Should we eliminate this flag and just reestimate the hashmap value every X records or so continuously? It can be non-linear probe as well.. 100, 1000, 10000, 100000 to amortize cost?

n3nash replied:
Yeah, I thought of that but I'm unsure of the performance of the ObjectSizeEstimator for such large objects, say a HashMap with 100K entries. Do you think we should performance test that now and do this continuously ?

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java:
```diff
@@ -33,88 +33,82 @@
+  // Value converter to convert value type to bytes
+  final private Converter<R> valueConverter;
+  // Find the actual estimated payload size after inserting N records
+  final private static int NUMBER_OF_RECORDS_TO_ESTIMATE_PAYLOAD_SIZE = 100;
```
done.

vinothchandar commented:
@n3nash we can move it back to IOHandle if that eases things. May be a static helper closer to the Map itself? since this math can be reused for compaction too? My original comment was on not having the a lot getter/setters for these spark props, which the localizing of variables addressed.

n3nash commented:
@vinothchandar Revised the PR. Unfortunately, cannot get spark defaults since they are hard-coded. I've added link to the places where they are hardcoded in the spark code in the comments. If there are more changes you propose, can we merge this and create a follow up ticket which I can take up in the following week ?
